File: frontend/src/api/priceList.ts
/**
 * ================================================================================
 * Price List API Client
 * ================================================================================
 * Version: 1.0.0
 * Created: 2025-12-11
 *
 * API client for Customer Price List management with date-based validity,
 * Excel import/export, and price resolution.
 * ================================================================================
 */

import apiClient from './client';

// ============================================================================
// TYPES
// ============================================================================

export interface PriceList {
    id: number;
    price_list_name: string;
    description?: string;
    valid_from: string; // YYYY-MM-DD
    valid_to?: string; // YYYY-MM-DD
    is_active: boolean;
    items_count: number;
    customers_count: number;
    status: 'active' | 'expired' | 'upcoming' | 'inactive';
    created_by?: string;
    created_at: string;
    updated_at: string;
}

export interface PriceListCreate {
    price_list_name: string;
    description?: string;
    valid_from: string; // YYYY-MM-DD
    valid_to?: string; // YYYY-MM-DD
    is_active?: boolean;
}

export interface PriceListUpdate {
    price_list_name?: string;
    description?: string;
    valid_from?: string;
    valid_to?: string;
    is_active?: boolean;
}

export interface PriceListItem {
    id: number;
    price_list_id: number;
    item_id: number;
    item_name: string;
    item_sku?: string;
    price: number;
    notes?: string;
    created_at: string;
    updated_at: string;
}

export interface PriceListItemCreate {
    item_id: number;
    price: number;
    notes?: string;
}

export interface PriceListListResponse {
    price_lists: PriceList[];
    total: number;
    page: number;
    limit: number;
}

export interface PriceListItemsResponse {
    items: PriceListItem[];
    total: number;
}

export interface ExcelImportResponse {
    success: boolean;
    items_imported: number;
    items_updated: number;
    items_failed: number;
    errors: Array<{
        row_number: number;
        sku: string;
        error: string;
    }>;
    message: string;
}

export interface DuplicatePriceListRequest {
    new_name: string;
    copy_items?: boolean;
    valid_from?: string;
    valid_to?: string;
}

export interface ResolvedPrice {
    customer_id: number;
    item_id: number;
    price: number;
    source: 'price_list' | 'zoho_default';
    price_list_id?: number;
    price_list_name?: string;
    date_resolved_for: string;
    is_price_list_active: boolean;
}

export interface CustomerPriceListInfo {
    customer_id: number;
    company_name: string;
    contact_name?: string;
    price_list_id?: number;
    price_list_name?: string;
    price_list_status?: string;
}

export interface AssignedCustomersResponse {
    price_list_id: number;
    price_list_name: string;
    customers: CustomerPriceListInfo[];
    total: number;
}

export interface PriceHistoryItem {
    id: number;
    price_list_id: number;
    item_id?: number;
    item_name?: string;
    field_changed: string;
    old_value?: string;
    new_value?: string;
    changed_by?: string;
    changed_at: string;
}

export interface PriceHistoryResponse {
    history: PriceHistoryItem[];
    total: number;
}

export interface PriceListStats {
    total_price_lists: number;
    active_price_lists: number;
    expired_price_lists: number;
    upcoming_price_lists: number;
    total_customers_with_price_lists: number;
    expiring_within_30_days: number;
}

// ============================================================================
// API CLIENT
// ============================================================================

export const priceListAPI = {
    /**
     * List all price lists with optional filters
     */
    list: async (params?: {
        status_filter?: 'active' | 'expired' | 'upcoming';
        date_filter?: string;
        page?: number;
        limit?: number;
    }): Promise<PriceListListResponse> => {
        const response = await apiClient.get('/price-lists', { params });
        return response.data;
    },

    /**
     * Create new price list
     */
    create: async (data: PriceListCreate): Promise<PriceList> => {
        const response = await apiClient.post('/price-lists', data);
        return response.data;
    },

    /**
     * Get price list by ID
     */
    getById: async (priceListId: number): Promise<PriceList> => {
        const response = await apiClient.get(`/price-lists/${priceListId}`);
        return response.data;
    },

    /**
     * Update price list
     */
    update: async (priceListId: number, data: PriceListUpdate): Promise<PriceList> => {
        const response = await apiClient.put(`/price-lists/${priceListId}`, data);
        return response.data;
    },

    /**
     * Delete price list
     */
    delete: async (priceListId: number): Promise<void> => {
        await apiClient.delete(`/price-lists/${priceListId}`);
    },

    /**
     * Duplicate price list
     */
    duplicate: async (
        priceListId: number,
        data: DuplicatePriceListRequest
    ): Promise<PriceList> => {
        const response = await apiClient.post(`/price-lists/${priceListId}/duplicate`, data);
        return response.data;
    },

    // ============================================================================
    // ITEMS
    // ============================================================================

    /**
     * Get all items in price list
     */
    getItems: async (priceListId: number): Promise<PriceListItemsResponse> => {
        const response = await apiClient.get(`/price-lists/${priceListId}/items`);
        return response.data;
    },

    /**
     * Add or update single item
     */
    addOrUpdateItem: async (
        priceListId: number,
        data: PriceListItemCreate
    ): Promise<PriceListItem> => {
        const response = await apiClient.post(`/price-lists/${priceListId}/items`, data);
        return response.data;
    },

    /**
     * Bulk add or update items
     */
    bulkAddOrUpdateItems: async (
        priceListId: number,
        items: PriceListItemCreate[]
    ): Promise<{ added: number; updated: number; errors: any[] }> => {
        const response = await apiClient.post(`/price-lists/${priceListId}/items/bulk`, {
            items,
        });
        return response.data;
    },

    /**
     * Update item price
     */
    updateItem: async (
        priceListId: number,
        itemId: number,
        data: PriceListItemCreate
    ): Promise<PriceListItem> => {
        const response = await apiClient.put(
            `/price-lists/${priceListId}/items/${itemId}`,
            data
        );
        return response.data;
    },

    /**
     * Remove item from price list
     */
    deleteItem: async (priceListId: number, itemId: number): Promise<void> => {
        await apiClient.delete(`/price-lists/${priceListId}/items/${itemId}`);
    },

    // ============================================================================
    // EXCEL IMPORT/EXPORT
    // ============================================================================

    /**
     * Download Excel template
     */
    downloadTemplate: async (): Promise<Blob> => {
        const response = await apiClient.get('/price-lists/template/download', {
            responseType: 'blob',
        });
        return response.data;
    },

    /**
     * Import from Excel
     */
    importFromExcel: async (
        priceListId: number,
        file: File
    ): Promise<ExcelImportResponse> => {
        const formData = new FormData();
        formData.append('file', file);

        const response = await apiClient.post(`/price-lists/${priceListId}/import`, formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });
        return response.data;
    },

    /**
     * Export to Excel
     */
    exportToExcel: async (priceListId: number): Promise<Blob> => {
        const response = await apiClient.get(`/price-lists/${priceListId}/export`, {
            responseType: 'blob',
        });
        return response.data;
    },

    // ============================================================================
    // CUSTOMER ASSIGNMENT
    // ============================================================================

    /**
     * Get assigned customers
     */
    getAssignedCustomers: async (
        priceListId: number
    ): Promise<AssignedCustomersResponse> => {
        const response = await apiClient.get(`/price-lists/${priceListId}/customers`);
        return response.data;
    },

    // ============================================================================
    // PRICE RESOLUTION
    // ============================================================================

    /**
     * Resolve price for customer + item combination
     */
    resolvePrice: async (
        customerId: number,
        itemId: number,
        dateFor?: string
    ): Promise<ResolvedPrice> => {
        const response = await apiClient.get(
            `/price-lists/resolve-price/customer/${customerId}/item/${itemId}`,
            {
                params: dateFor ? { date_for: dateFor } : undefined,
            }
        );
        return response.data;
    },

    // ============================================================================
    // HISTORY & STATS
    // ============================================================================

    /**
     * Get price list change history
     */
    getHistory: async (priceListId: number, limit?: number): Promise<PriceHistoryResponse> => {
        const response = await apiClient.get(`/price-lists/${priceListId}/history`, {
            params: limit ? { limit } : undefined,
        });
        return response.data;
    },

    /**
     * Get statistics
     */
    getStats: async (): Promise<PriceListStats> => {
        const response = await apiClient.get('/price-lists/stats/summary');
        return response.data;
    },
};

export default priceListAPI;

File: backend/app/schemas/price_list.py
"""
================================================================================
Marketplace ERP - Customer Price List Schemas
================================================================================
Version: 1.0.0
Created: 2025-12-11

Pydantic models for customer price list management, including price lists,
items, history tracking, and Excel import/export.
================================================================================
"""

from pydantic import BaseModel, Field, validator, field_serializer
from typing import Optional, List, Dict, Any
from datetime import date, datetime
from decimal import Decimal


# ============================================================================
# PRICE LIST SCHEMAS
# ============================================================================

class PriceListBase(BaseModel):
    """Base price list fields"""
    price_list_name: str = Field(..., min_length=1, max_length=255, description="Price list name")
    description: Optional[str] = Field(None, description="Optional description")
    valid_from: date = Field(..., description="Date from which price list is active")
    valid_to: Optional[date] = Field(None, description="Date until which price list is active (NULL = indefinite)")
    is_active: bool = Field(True, description="Manual override to enable/disable")

    @validator('valid_to')
    def validate_date_range(cls, v, values):
        """Ensure valid_to is after valid_from"""
        if v and 'valid_from' in values and v < values['valid_from']:
            raise ValueError('valid_to must be on or after valid_from')
        return v


class PriceListCreate(PriceListBase):
    """Create price list request"""
    pass


class PriceListUpdate(BaseModel):
    """Update price list request (all fields optional)"""
    price_list_name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    valid_from: Optional[date] = None
    valid_to: Optional[date] = None
    is_active: Optional[bool] = None


class PriceListResponse(PriceListBase):
    """Price list response with metadata"""
    id: int
    items_count: int = Field(0, description="Number of items in price list")
    customers_count: int = Field(0, description="Number of customers assigned")
    status: str = Field(..., description="active | expired | upcoming")
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
    
    @field_serializer('created_by')
    def serialize_created_by(self, value):
        """Convert UUID to string"""
        return str(value) if value else None


class PriceListListResponse(BaseModel):
    """Paginated price list list"""
    price_lists: List[PriceListResponse]
    total: int
    page: int
    limit: int


# ============================================================================
# PRICE LIST ITEM SCHEMAS
# ============================================================================

class PriceListItemBase(BaseModel):
    """Base price list item fields"""
    item_id: int = Field(..., description="Zoho item ID")
    price: Decimal = Field(..., gt=0, max_digits=10, decimal_places=2, description="Item price in INR")
    notes: Optional[str] = Field(None, description="Optional notes about this price")


class PriceListItemCreate(PriceListItemBase):
    """Create/update price list item"""
    pass


class PriceListItemResponse(PriceListItemBase):
    """Price list item response with item details"""
    id: int
    price_list_id: int
    item_name: str = Field(..., description="Item name from zoho_item_master")
    item_sku: Optional[str] = Field(None, description="Item SKU")
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PriceListItemsResponse(BaseModel):
    """List of price list items"""
    items: List[PriceListItemResponse]
    total: int


class BulkPriceListItemCreate(BaseModel):
    """Bulk create/update price list items""" 
    items: List[PriceListItemCreate] = Field(..., description="List of items to add/update")


# ============================================================================
# PRICE HISTORY SCHEMAS
# ============================================================================

class PriceHistoryItem(BaseModel):
    """Price history entry"""
    id: int
    price_list_id: int
    item_id: Optional[int] = None
    item_name: Optional[str] = None
    field_changed: str
    old_value: Optional[str] = None
    new_value: Optional[str] = None
    changed_by: Optional[str] = None
    changed_at: datetime

    class Config:
        from_attributes = True


class PriceHistoryResponse(BaseModel):
    """Price history list"""
    history: List[PriceHistoryItem]
    total: int


# ============================================================================
# CUSTOMER ASSIGNMENT SCHEMAS
# ============================================================================

class CustomerPriceListInfo(BaseModel):
    """Customer with price list assignment"""
    customer_id: int
    company_name: str
    contact_name: Optional[str] = None
    price_list_id: Optional[int] = None
    price_list_name: Optional[str] = None
    price_list_status: Optional[str] = None  # active | expired | upcoming


class AssignedCustomersResponse(BaseModel):
    """Customers assigned to a price list"""
    price_list_id: int
    price_list_name: str
    customers: List[CustomerPriceListInfo]
    total: int


# ============================================================================
# PRICE RESOLUTION SCHEMAS
# ============================================================================

class ResolvedPriceResponse(BaseModel):
    """Resolved price for customer + item combination"""
    customer_id: int
    item_id: int
    price: Decimal = Field(..., max_digits=10, decimal_places=2)
    source: str = Field(..., description="price_list | zoho_default")
    price_list_id: Optional[int] = None
    price_list_name: Optional[str] = None
    date_resolved_for: date = Field(..., description="Date for which price was resolved")
    is_price_list_active: bool = Field(..., description="Whether price list is active for this date")


# ============================================================================
# EXCEL IMPORT/EXPORT SCHEMAS
# ============================================================================

class ExcelImportRow(BaseModel):
    """Single row from Excel import"""
    sku: str = Field(..., description="Item SKU")
    price: Decimal = Field(..., gt=0)
    notes: Optional[str] = None


class ExcelImportRequest(BaseModel):
    """Excel import data"""
    rows: List[ExcelImportRow]


class ExcelImportError(BaseModel):
    """Import error details"""
    row_number: int
    sku: str
    error: str


class ExcelImportResponse(BaseModel):
    """Excel import result"""
    success: bool
    items_imported: int
    items_updated: int
    items_failed: int
    errors: List[ExcelImportError] = []
    message: str


# ============================================================================
# DUPLICATE PRICE LIST SCHEMAS
# ============================================================================

class DuplicatePriceListRequest(BaseModel):
    """Duplicate price list request"""
    new_name: str = Field(..., min_length=1, max_length=255, description="Name for duplicated price list")
    copy_items: bool = Field(True, description="Whether to copy all items")
    valid_from: Optional[date] = Field(None, description="Valid from date for new list (required if copying)")
    valid_to: Optional[date] = Field(None, description="Valid to date for new list")


# ============================================================================
# STATISTICS SCHEMAS
# ============================================================================

class PriceListStatsResponse(BaseModel):
    """Price list statistics"""
    total_price_lists: int
    active_price_lists: int
    expired_price_lists: int
    upcoming_price_lists: int
    total_customers_with_price_lists: int
    expiring_within_30_days: int
